TELCODOCS-1085 - custom source CRs for PolicyGenTemplate CR (#60970)
openshift/openshift-docs: modules/ztp-adding-new-content-to-gitops-ztp.adoc

// * scalability_and_performance/ztp_far_edge/ztp-advanced-policy-config.adoc

:_content-type: PROCEDURE

###################################################
[id="ztp-adding-new-content-to-gitops-ztp_{context}"]

> = Adding custom content to the {ztp} pipeline

Reviewer: Instead of adding this procedure to the `modules/ztp-adding-new-content-to-gitops-ztp.adoc` module, suggest creating a new module, `modules/ztp-adding-custom-crs-to-gitops-ztp.adoc`

Author: The doc story suggested that the work done in [CNF-6925](https://issues.redhat.com/browse/CNF-6925) likely obsoletes the procedure in [Adding new content to the GitOps ZTP pipeline](https://docs.openshift.com/container-platform/4.11/scalability_and_performance/ztp_far_edge/ztp-advanced-policy-config.html#ztp-adding-new-content-to-gitops-ztp_ztp-advanced-policy-config)

The suggestion was to rewrite the existing module, preserving the file name/id for backwards compatibility. 
SME supported that suggestion so I've gone with that.

> 

Reviewer: Need some prereqs here to set the context for the procedure, eg., you must have a site config repo, ArgoCD application configured, etc.

Reviewer 2: I am not getting the reason behind to set pre-requisites here as there is no change in implementation design. we are proposing another way how to source-crs with customization as per end-users requirements.

Author: ok, we'll leave as is so.

Perform the following procedure to add new content to the {ztp} pipeline.

.Procedure

> . Create a subdirectory,  `source-crs`,  in the directory containing the `kustomization.yaml` for the `PolicyGenTemplate` CR. 

Reviewer: Generally, you use multiple PGT CRs to manage various aspects of a fleet of clusters.

. Add your custom CRs to this folder, for example:
+
[source,text]
----
example

> ├── policygentemplates

Reviewer: ```suggestion
└── policygentemplates
```

Author: ok

    ├── dev.yaml

>     ├── kustomization.yaml│   

Reviewer: ```suggestion
    ├── kustomization.yaml
```

    ├── mec-edge-sno1.yaml
    ├── sno.yaml
    └── source-crs <1>
        ├── PaoCatalogSource.yaml
        ├── PaoSubscription.yaml
        ├── bell
        |   ├── apiserver-config.yaml
        |   └── disable-nic-lldp.yaml
        └── elasticsearch
            ├── ElasticsearchNS.yaml
            └── ElasticsearchOperatorGroup.yaml
----
<1> The `sources-crs` subdirectory is in the same directory as `kustomization.yaml`.

> . Update the `PolicyGenTemplate` to include references to the content you added to your `custom-crs` directory.

Reviewer: ```suggestion
. Update the required `PolicyGenTemplate` CRs to include references to the content you added in the `source-crs` directory.
```

Author: done

      remediationAction: inform 
      policyName: "group-dev-lso-sub" 
    # Performance Addon Operator
    - fileName: PaoSubscriptionNS.yaml
      remediationAction: inform 
      policyName: "group-dev-pao-ns" 
    - fileName: PaoSubscriptionCatalogSource.yaml
      remediationAction: inform 
      policyName: "group-dev-pao-cat-source" 
      spec:
        image: <image URL here>
    - fileName: PaoSubscription.yaml
      remediationAction: inform 
      policyName: "group-dev-pao-sub" 
    #These are custom local polices that come from the source-crs directory in the git repo
    #Elasticsearch Operator
    - fileName: elasticsearch/ElasticsearchNS.yaml <1>
      remediationAction: inform 
      policyName: "group-dev-elasticsearch-ns" 
    - fileName: elasticsearch/ElasticsearchOperatorGroup.yaml
      remediationAction: inform 
      policyName: "group-dev-elasticsearch-operator-group" 
    #Bell Custom Resources
    - fileName: bell/apiserver-config.yaml
      remediationAction: inform 
      policyName: "group-dev-apiserver-config" 
    - fileName: bell/disable-nic-lldp.yaml
      remediationAction: inform 
      policyName: "group-dev-disable-nic-lldp" 
----

> <1> The `fileName` includes the sub-directory within the `source-crs` directory.

Reviewer: ```suggestion
<1> Set `fileName` to include the custom CR subfolder from the `/source-crs` parent folder.
```

Author: done

. Commit the `PolicyGenTemplate` change in Git, and then push to the Git repository being monitored by the GitOps ZTP Argo CD application.

> 

Reviewer: Are there verification steps that we can include to tell the user how to verify that the custom CRs were successfully applied?

Author: Updated to include verification steps

    - fileName: PaoSubscriptionNS.yaml
      remediationAction: inform 
      policyName: "group-dev-pao-ns" 
    - fileName: PaoSubscriptionCatalogSource.yaml
      remediationAction: inform 
      policyName: "group-dev-pao-cat-source" 
      spec:
        image: <image URL here>
    - fileName: PaoSubscription.yaml
      remediationAction: inform 
      policyName: "group-dev-pao-sub" 
    #These are custom local polices that come from the source-crs directory in the git repo
    #Elasticsearch Operator
    - fileName: elasticsearch/ElasticsearchNS.yaml <1>
      remediationAction: inform 
      policyName: "group-dev-elasticsearch-ns" 
    - fileName: elasticsearch/ElasticsearchOperatorGroup.yaml
      remediationAction: inform 
      policyName: "group-dev-elasticsearch-operator-group" 
    #Bell Custom Resources
    - fileName: bell/apiserver-config.yaml
      remediationAction: inform 
      policyName: "group-dev-apiserver-config" 
    - fileName: bell/disable-nic-lldp.yaml
      remediationAction: inform 
      policyName: "group-dev-disable-nic-lldp" 
----
<1> Set `fileName` to include the custom CR subfolder from the `/source-crs` parent.

> . Commit the `PolicyGenTemplate` change in Git, and then push to the Git repository being monitored by the GitOps ZTP Argo CD application.

Reviewer 2: GitOps ZTP Argo CD application => GitOps ZTP ArgoCD policies application - Good to update this way.

> . Update the `ClusterGroupUpdate` CR to include the changed `PolicyGenTemplate` and save it as `cgu-test.yaml`.

Reviewer 2: `ClusterGroupUpgrade`

Author: opps - corrected. Thanks

+
[source,yaml]
----
apiVersion: ran.openshift.io/v1alpha1
kind: ClusterGroupUpgrade
metadata:
  name: custom-source-cr

>   namespace: ztp-group

Reviewer 2: NS should be ztp-clusters based on your above sample example.

spec:
  managedPolicies:

>     - group-du-sno-v4.14-config-policy

Reviewer: Is this correct? the example above has a different policy name.

Reviewer 2: should match with your above sample example.
Hence, should be as "group-dev-config-policy"

Author: ok

  enable: true
  clusters:
  - cluster1
  remediationStrategy:
    maxConcurrency: 2
    timeout: 240
----

> . Apply the updated `ClusterGroupUpdate` CR by running the following command:

Reviewer 2: ClusterGroupUpgrade

Author: ok

[source, terminal]
+
----
$ oc apply -f cgu-test.yaml
----

.Verification

. Check that the updates have succeeded by running the following command:
+
[source, terminal]
----
$ oc  get cgu -A
----
+
.Example output
+
[source, terminal]
----
NAMESPACE     NAME               AGE   STATE        DETAILS

> ztp-group     custom-source-cr   6s    InProgress   Remediating non-compliant policies

Reviewer 2: NS should be ztp-clusters based on your previous example.

Author: ok and thanks.

// * scalability_and_performance/ztp_far_edge/ztp-advanced-policy-config.adoc

:_content-type: PROCEDURE

[id="ztp-adding-new-content-to-gitops-ztp_{context}"]
= Adding custom content to the {ztp} pipeline

Perform the following procedure to add new content to the {ztp} pipeline.

.Procedure

> . Create a subdirectory,  `source-crs`,  in the directory containing the `kustomization.yaml` for the `PolicyGenTemplate` CR. 

Author: ok

> . Add your custom CRs to this folder, for example:

Author: ok

+
[source,text]
----
example
└── policygentemplates
    ├── dev.yaml
    ├── kustomization.yaml
    ├── mec-edge-sno1.yaml
    ├── sno.yaml
    └── source-crs <1>
        ├── PaoCatalogSource.yaml
        ├── PaoSubscription.yaml

>         ├── bell

Author: Thanks, now addressed in PR https://github.com/openshift/openshift-docs/pull/64949

        |   ├── apiserver-config.yaml
        |   └── disable-nic-lldp.yaml
        └── elasticsearch
            ├── ElasticsearchNS.yaml
            └── ElasticsearchOperatorGroup.yaml
----

> <1> The `sources-crs` subdirectory is in the same directory as `kustomization.yaml`.

Author: ok

> <1> The `sources-crs` subdirectory is in the same directory as `kustomization.yaml`.

Author: Thanks for the review @lahinson

> <1> The `sources-crs` subdirectory is in the same directory as the `kustomization.yaml` file.

Reviewer 3: source-crs

Author: corrected, thanks

Author: PR https://github.com/openshift/openshift-docs/pull/64949

> . Update the required `PolicyGenTemplate` CRs to include references to the content you added in the `custom-crs` directory.

Reviewer 3: There is no custom-crs directory here, either in the custom directory under source-crs or mention source-crs/custom directory.

Author: ok

      remediationAction: inform 
      policyName: "group-dev-pao-cat-source" 
      spec:
        image: <image URL here>
    - fileName: PaoSubscription.yaml
      remediationAction: inform 
      policyName: "group-dev-pao-sub" 
    #These are custom local polices that come from the source-crs directory in the git repo
    #Elasticsearch Operator
    - fileName: elasticsearch/ElasticsearchNS.yaml <1>
      remediationAction: inform 
      policyName: "group-dev-elasticsearch-ns" 
    - fileName: elasticsearch/ElasticsearchOperatorGroup.yaml
      remediationAction: inform 
      policyName: "group-dev-elasticsearch-operator-group" 
    #Bell Custom Resources
    - fileName: bell/apiserver-config.yaml
      remediationAction: inform 
      policyName: "group-dev-apiserver-config" 
    - fileName: bell/disable-nic-lldp.yaml
      remediationAction: inform 
      policyName: "group-dev-disable-nic-lldp" 
----
<1> Set `fileName` to include the custom CR subfolder from the `/source-crs` parent.

. Commit the `PolicyGenTemplate` change in Git, and then push to the Git repository being monitored by the GitOps ZTP Argo CD policies application.

> . Update the `ClusterGroupUpgrade` CR to include the changed `PolicyGenTemplate` and save it as `cgu-test.yaml`.

Reviewer 4: ```suggestion
. Update the `ClusterGroupUpgrade` CR to include the changed `PolicyGenTemplate` and save it as `cgu-test.yaml`, as shown in the following example:
```
- Nit: Consider add introductory phrase to the block of sample code followed by a colon. If you want to make the change, you can add this to the Jira for fixing the alignment of the example output.

Author: Thanks, Changed this in new PR
https://github.com/openshift/openshift-docs/pull/64949

+
[source,yaml]
----
apiVersion: ran.openshift.io/v1alpha1
kind: ClusterGroupUpgrade
metadata:
  name: custom-source-cr
  namespace: ztp-clusters
spec:
  managedPolicies:
    - group-dev-config-policy
  enable: true
  clusters:
  - cluster1
  remediationStrategy:
    maxConcurrency: 2
    timeout: 240
----

. Apply the updated `ClusterGroupUpgrade` CR by running the following command:
[source, terminal]
+
----
$ oc apply -f cgu-test.yaml
----

.Verification

. Check that the updates have succeeded by running the following command:
+
[source, terminal]
----
$ oc  get cgu -A
----
+
.Example output
+
[source, terminal]
----

> NAMESPACE     NAME               AGE   STATE        DETAILS

Reviewer: Nit: the example output is mis-aligned in the HTML

Reviewer 4: Aidan is correct. The sample output is misaligned. It's not a big enough issue to prevent a merge, but please write up a small Jira to correct the typo. Thanks!

Author: Corrected this in new PR https://github.com/openshift/openshift-docs/pull/64949
